File: built-in plugins dsp/mb-comp-opto.cpp
/**
 * MB Opto Comp
 * Category : effect
 * Type     : compressor
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Smooth optical compressor
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_COMP_OPTO_H
#define MB_COMP_OPTO_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbCompOpto : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-comp-opto";
    static constexpr const char* PLUGIN_NAME    = "MB Opto Comp";
    static constexpr const char* PLUGIN_TYPE    = "compressor";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = -24f;  // range [-60, 0]
    float ratio = 3f;  // range [1, 10]
    float attack = 20f;  // range [5, 100]
    float release = 200f;  // range [50, 2000]
    };

    MbCompOpto() = default;
    ~MbCompOpto() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, -60f, 0f);
        params.ratio = std::clamp(params.ratio, 1f, 10f);
        params.attack = std::clamp(params.attack, 5f, 100f);
        params.release = std::clamp(params.release, 50f, 2000f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Opto Comp
        return input;
    }
};

#endif // MB_COMP_OPTO_H

File: built-in plugins dsp/mb-comp-opto.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbCompOptoPlugin: PluginDefinition = {
  id: "mb-comp-opto",
  slug: "mb-comp-opto",
  name: "MB Opto Comp",
  category: "effect",
  type: "compressor",
  version: "1.0.0",
  description: "Smooth optical compressor",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "threshold",
      name: "Threshold",
      type: "float",
      defaultValue: -24,
      minValue: -60,
      maxValue: 0,
      automatable: true,
    },
    {
      id: "ratio",
      name: "Ratio",
      type: "float",
      defaultValue: 3,
      minValue: 1,
      maxValue: 10,
      automatable: true,
    },
    {
      id: "attack",
      name: "Attack",
      type: "float",
      defaultValue: 20,
      minValue: 5,
      maxValue: 100,
      automatable: true,
    },
    {
      id: "release",
      name: "Release",
      type: "float",
      defaultValue: 200,
      minValue: 50,
      maxValue: 2000,
      automatable: true,
    },
  ],
  defaultPreset: { threshold: -24, ratio: 3, attack: 20, release: 200 },
};

export default MbCompOptoPlugin;
